In [39]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
import scipy

# Data Cleaning, Missing Data and Normalization

## Removing Non-Predictive Features

In [40]:
# Load the data
df_unclean = pd.read_csv('CommViolPredUnnormalizedData.csv', na_values=['?'])


In [41]:
# Removing the non-predictive features
columns_to_remove = [
    'communityname',
    'state',
    'countyCode',
    'communityCode',
    'fold',
    'Unnamed: 0'
]

# Removing potential target features
columns_to_remove += [
        'murders', 
        'murdPerPop', 
        'rapes', 
        'rapesPerPop', 
        'robberies', 
        'robbbPerPop', 
        'assaults', 
        'burglaries',
        'burglPerPop', 
        'larcenies', 
        'larcPerPop',
        'autoTheft', 
        'autoTheftPerPop', 
        'arsons', 
        'arsonsPerPop',
        'ViolentCrimesPerPop', 
        'nonViolPerPop'
    ]


df_cleaned = df_unclean.drop(columns=columns_to_remove)
df_cleaned

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0          11980           3.10          1.37         91.78          6.50   
1          23123           2.82          0.80         95.57          3.44   
2          29344           2.43          0.74         94.33          3.43   
3          16656           2.40          1.70         97.35          0.50   
4          11245           2.76          0.53         89.16          1.17   
...          ...            ...           ...           ...           ...   
2210       56216           3.07          6.87         61.68         15.23   
2211       12251           2.68         21.18         76.65          1.52   
2212       32824           2.46          0.52         92.62          0.98   
2213       13547           2.89          3.37         69.91          0.90   
2214       28898           2.61          2.39         71.27          9.09   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  \
0            1.88        12.47        21.44        10.93       11.33  ...   
1            0.85        11.01        21.30        10.48       17.18  ...   
2            2.35        11.36        25.88        11.01       10.28  ...   
3            0.70        12.55        25.20        12.19       17.57  ...   
4            0.52        24.46        40.53        28.69       12.65  ...   
...           ...          ...          ...          ...         ...  ...   
2210        29.86        15.46        30.16        14.34        8.08  ...   
2211         1.29        17.36        31.23        16.97       12.57  ...   
2212        11.00        11.81        20.96         9.53       20.73  ...   
2213        62.11        17.16        30.01        14.73       10.42  ...   
2214        24.43        12.99        25.21        11.63       12.12  ...   

      LandArea  PopDens  PctUsePubTrans  PolicCars  PolicOperBudg  \
0          6.5   1845.9            9.63        NaN            NaN   
1         10.6   2186.7            3.84        NaN            NaN   
2         10.6   2780.9            4.37        NaN            NaN   
3          5.2   3217.7            3.31        NaN            NaN   
4         11.5    974.2            0.38        NaN            NaN   
...        ...      ...             ...        ...            ...   
2210      16.7   3365.4            0.59        NaN            NaN   
2211       7.3   1682.8            1.15        NaN            NaN   
2212      27.5   1195.2            0.12        NaN            NaN   
2213       6.3   2142.2            0.00        NaN            NaN   
2214      21.7   1331.0            1.39        NaN            NaN   

      LemasPctPolicOnPatr  LemasGangUnitDeploy  LemasPctOfficDrugUn  \
0                     NaN                  NaN                  0.0   
1                     NaN                  NaN                  0.0   
2                     NaN                  NaN                  0.0   
3                     NaN                  NaN                  0.0   
4                     NaN                  NaN                  0.0   
...                   ...                  ...                  ...   
2210                  NaN                  NaN                  0.0   
2211                  NaN                  NaN                  0.0   
2212                  NaN                  NaN                  0.0   
2213                  NaN                  NaN                  0.0   
2214                  NaN                  NaN                  0.0   

      PolicBudgPerPop  assaultPerPop  
0                 NaN          32.81  
1                 NaN         102.05  
2                 NaN          38.74  
3                 NaN         190.93  
4                 NaN         112.14  
...               ...            ...  
2210              NaN         280.29  
2211              NaN          82.73  
2212              NaN         261.29  
2213              NaN         620.48  
2214              NaN         502.41  

[2215

## Handling Features with Missing Data

First, the missing data in the dataset is identified. For each feature  with misisng data, the percentage of missing data is calculated to help decide on how to handle the missing data.

In [42]:

# Calculate missing values statistics
total_cells = np.prod(df_cleaned.shape)
total_missing = df_cleaned.isnull().sum().sum()
        

missing_values = df_cleaned.isnull().sum()
missing_percentages = (missing_values / len(df_cleaned)) * 100
        
# Create summary DataFrame
missing_info = pd.DataFrame({'Missing Values': missing_values, 'Percentage Missing': missing_percentages})
missing_info = missing_info[missing_info['Missing Values'] > 0].sort_values('Percentage Missing', ascending=False)

# Display missing values statistics
if not missing_info.empty:
    print("-" * 50)
    for idx, row in missing_info.iterrows():
        print(f"{idx}:")
        print(f"  Missing values: {row['Missing Values']:,}")
        print(f"  Percentage missing: {row['Percentage Missing']:.2f}%")
        print("-" * 50)
else:
    print("\nNo missing values found in the dataset!")

--------------------------------------------------
PctPolicHisp:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
PolicBudgPerPop:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
LemasPctPolicOnPatr:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
PolicOperBudg:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
PolicCars:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
PolicAveOTWorked:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
NumKindsDrugsSeiz:
  Missing values: 1,872.0
  Percentage missing: 84.51%
--------------------------------------------------
OfficAssgnDrugUnits:
  Missing values: 1,872.0
  Percentage missing: 84.51%
----------------

The features which have 84.51% missing data are removed from the dataset since the high percentage of missing data makes it difficult to impute the missing data.

In [43]:
# Drop rows with high percentage of missing values
for idx, row in missing_info.iterrows():
    if row['Percentage Missing'] > 50:
        df_cleaned = df_cleaned.drop(columns=idx)

For the target the `assaultPerPop` feature, the rows with missing data are removed from the dataset since this is the target feature and therefore imputing the missing values could lead to biased results.

In [44]:
# Drop rows with missing values
df_cleaned = df_cleaned.dropna(subset=['assaultPerPop'])

For the `OtherPerCap` feature, the missing values are imputed with the median of the feature as there is only one missing value. The median is chosen as there are outliers in the feature and a large standard deviation which makes the mean sensitive to outliers. 

In [45]:
# Ctatistics for OtherPerCap
print(df_cleaned['OtherPerCap'].describe())

# Calculate the lower and upper bounds for outliers
Q1 = df_cleaned['OtherPerCap'].quantile(0.25)
Q3 = df_cleaned['OtherPerCap'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Find outliers
outliers = df_cleaned[(df_cleaned['OtherPerCap'] < lower_bound) | 
                      (df_cleaned['OtherPerCap'] > upper_bound)]['OtherPerCap']

print(f"Number of outliers: {len(outliers)}")

count      2201.000000
mean       9453.911404
std        7932.287114
min           0.000000
25%        5529.000000
50%        8222.000000
75%       11535.000000
max      137000.000000
Name: OtherPerCap, dtype: float64
Number of outliers: 119


In [46]:
# Impute missing values with the median
df_cleaned.loc[:, 'OtherPerCap'] = df_cleaned['OtherPerCap'].fillna(df_cleaned['OtherPerCap'].median())